akka-http2-support/src/main/scala/akka/http/impl/engine/http2/PriorityTree.scala

```diff
@@ -0,0 +1,154 @@
+}
+
+/** INTERNAL API */
+private[http2] trait PriorityTree {
+  /**
+   * Returns a new priority tree containing the new or existing and updated stream.
+   */
+  def insertOrUpdate(streamId: Int, streamDependency: Int, weight: Int, exclusive: Boolean): PriorityTree
+
+  /**
+   * Returns the root node of the tree.
+   */
+  def rootNode: PriorityNode
+
+  /**
+   * Gives an ASCII representation of the tree.
+   */
+  def print: String
+}
+
+/** INTERNAL API */
+private[http2] object PriorityTree {
+  def apply(): PriorityTree = create(TreeMap(0 → RootNode))
+
+  private final val DefaultWeight = 16
+
+  private val RootNode = PriorityInfo(0, 0, DefaultWeight, TreeSet.empty)
+  private def create(nodes: TreeMap[Int, PriorityInfo]): PriorityTreeImpl =
+    new PriorityTreeImpl(nodes)
+
+  private class PriorityTreeImpl(nodes: TreeMap[Int, PriorityInfo]) extends PriorityTree {
+    def rootNode: PriorityNode = node(0)
+
+    def insertOrUpdate(streamId: Int, streamDependency: Int, weight: Int, exclusive: Boolean): PriorityTree =
+      if (nodes.isDefinedAt(streamId)) update(streamId, streamDependency, weight, exclusive)
+      else insert(streamId, streamDependency, weight, exclusive)
+
+    private def insert(streamId: Int, streamDependency: Int, weight: Int, exclusive: Boolean): PriorityTreeImpl = {
+      require(!nodes.isDefinedAt(streamId))
+      require(streamId != streamDependency)
+
+      if (nodes.isDefinedAt(streamDependency)) {
+        val dependencyInfo = nodes(streamDependency)
+        if (!exclusive)
+          insertNode(PriorityInfo(streamId, streamDependency, weight, TreeSet.empty))
+            .updateNode(streamDependency)(updateChildren(_ + streamId))
+        else
+          insertNode(PriorityInfo(streamId, streamDependency, weight, dependencyInfo.childrenIds))
+            .updateNode(streamDependency)(updateChildren(_ ⇒ TreeSet(streamId)))
+      } else
+        insertNode(PriorityInfo(streamDependency, 0, DefaultWeight, TreeSet.empty)) // try again after creating intermediate
+          .insert(streamId, streamDependency, weight, exclusive)
+    }
+    private def update(streamId: Int, newStreamDependency: Int, newWeight: Int, newlyExclusive: Boolean): PriorityTree = {
+      require(nodes.isDefinedAt(streamId))
+      require(streamId != newStreamDependency)
+
+      if (nodes.isDefinedAt(newStreamDependency)) {
+        val oldInfo = nodes(streamId)
+
+        if (!newlyExclusive && newStreamDependency == dependencyOf(streamId)) // no dependency changes -> simple
+          updateNode(streamId)(_.copy(weight = newWeight))
+        else if (!dependsTransitivelyOn(newStreamDependency, streamId)) {
+          remove(streamId)
+            .insert(streamId, newStreamDependency, newWeight, newlyExclusive)
+            .updateNode(streamId)(updateChildren(newChildren ⇒ newChildren ++ oldInfo.childrenIds))
+        } else {
+          this // FIXME: actually implement moving nodes down in a dependency chain
```

Comment:
ok, not critical to have this up front - perhaps ticket for it?
This is about when we "reprioritise" under the same part of the tree right?